consensus/src/block_storage/sync_manager.rs

```diff
@@ -259,23 +298,29 @@ impl BlockRetriever {
     /// leader to drive quorum certificate creation The other peers from the quorum certificate
     /// will be randomly tried next.  If all members of the quorum certificate are exhausted, an
     /// error is returned
-    async fn retrieve_block_for_qc<'a>(
-        &'a mut self,
-        qc: &'a QuorumCert,
+    async fn retrieve_block_for_id(
+        &mut self,
+        block_id: HashValue,
+        peers: &mut Vec<&AccountAddress>,
         num_blocks: u64,
     ) -> anyhow::Result<Vec<Block>> {
-        let block_id = qc.certified_block().id();
-        let mut peers: Vec<&AccountAddress> = qc.ledger_info().signatures().keys().collect();
         let mut attempt = 0_u32;
-        loop {
-            if peers.is_empty() {
-                bail!(
-                    "Failed to fetch block {} in {} attempts: no more peers available",
-                    block_id,
-                    attempt
-                );
-            }
-            let peer = self.pick_peer(attempt, &mut peers);
+        let mut progress = 0;
+        let mut last_block_id = block_id;
+        let mut result_blocks: Vec<Block> = vec![];
+        let mut retrieve_batch_size = min(num_blocks, MAX_BLOCKS_PER_REQUEST);
+        if peers.is_empty() {
+            bail!(
+                "Failed to fetch block {} in {} attempts: no more peers available",
+                block_id,
+                attempt
+            );
+        }
+        let mut peer = self.pick_peer(attempt, peers);
+        while progress < num_blocks {
```

Comment:
we should split the function into 1. split into chunks 2. fetch a single chunk

Reply:
we use `self.network.request_block` to fetch a single chunk..?